### PR TITLE
[SPARK-48421][SQL] SPJ: Add documentation

### PR DESCRIPTION
**Pull request by szehon-ho**
### What changes were proposed in this pull request?
Add docs for SPJ

### Why are the changes needed?
There are no docs describing SPJ, even though it is mentioned in migration notes:  https://github.com/apache/spark/pull/46673

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Checked the new text

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

File: docs/sql-performance-tuning.md [hunk 1]
```diff
@@ -428,3 +428,101 @@ You can control the details of how AQE works by providing your own cost evaluato
       <td>3.2.0</td>
     </tr>
   </table>
+
+## Storage Partition Join
```
Comment (HyukjinKwon) [inline docs/sql-performance-tuning.md:432]: Can we have at least one working example? e.g., how do we use it with Parquet DSv2?

Comment (szehon-ho) [reply docs/sql-performance-tuning.md:432]: Thanks for review.  The mechanism works today if DSV2 partitioning implements KeyGroupedPartitioning. 
 Unfortunately the initial SPJ implementation did not support it for ParquetDSV2, so I think there's no data source in Spark code that we can have an example at the moment (the tests use the InMemoryTable which implements it).  I can potentially investigate implementing for ParquetDSV2 later.

Comment (HyukjinKwon) [reply docs/sql-performance-tuning.md:432]: Could we have an example with thridparty library that is open source?

Comment (szehon-ho) [reply docs/sql-performance-tuning.md:432]: @HyukjinKwon sure, I put iceberg example, let me know what you think

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
But would be great if we have eyes from other folks like @sunchao @cloud-fan

File: docs/sql-performance-tuning.md [hunk 2]
```diff
@@ -428,3 +428,118 @@ You can control the details of how AQE works by providing your own cost evaluato
...
+
+Storage Partition Join (SPJ) is an optimization technique in Spark SQL that makes use the existing storage layout to avoid the shuffle phase.
+
+This is a generalization of the concept of Bucket Joins, which is only applicable for [bucketed](sql-data-sources-load-save-functions.html#bucketing-sorting-and-partitioning) tables, to tables partitioned by functions registered in FunctionCatalog. Storage Partition Joins are currently supported for compatible V2 DataSources.
```
Comment (sunchao) [inline docs/sql-performance-tuning.md:436]: nit: this `bucketed` link doesn't work

File: docs/sql-performance-tuning.md [hunk 3]
```diff
@@ -428,3 +428,118 @@ You can control the details of how AQE works by providing your own cost evaluato
...
+
+The following SQL properties enable Storage Partition Join.
```
Comment (sunchao) [inline docs/sql-performance-tuning.md:438]: nit: perhaps `The following SQL properties enable Storage Partition Join and various optimizations of it`?

**Review by sunchao — APPROVED** [review]
LGTM too with a few nits, thanks @szehon-ho !

File: docs/sql-performance-tuning.md (hunk 2 above)
Comment (szehon-ho) [reply docs/sql-performance-tuning.md:436]: Hm, i built the site and it seems to work, but let me know if another one is better here?

File: docs/sql-performance-tuning.md (hunk 3 above)
Comment (szehon-ho) [reply docs/sql-performance-tuning.md]: done, added 'in different join queries with various optimizations.'  (as some flags are about different scenarios)

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by szehon-ho** [comment]
Thank you @HyukjinKwon @sunchao

**Pull request closed without merging**